visualization/app/codeCharta/ui/searchPanelModeSelector/searchPanelModeSelector.po.ts

```diff
@@ -1,17 +1,23 @@
-import { Page } from "puppeteer"
-import { delay } from "../../../puppeteer.helper"
-
 export class SearchPanelModeSelectorPageObject {
-	constructor(private page: Page) {}
-
 	public async toggleTreeView() {
-		await this.page.click("#tree-view")
-		await delay(500)
+		const wasOpen = await this.isTreeViewOpen()
+
+		await expect(page).toClick("#tree-view", { timeout: 3000 })
+
+		if (wasOpen) {
+			await page.waitForSelector("#search-panel-card", { visible: false })
+		} else {
+			await page.waitForSelector("#search-panel-card.expanded")
+		}
```

Comment:
I still suggest to return `!wasOpen`. That way the test has to do less work.

Right now there's a test that looks like:

```js
expect(await searchPanelModeSelector.isTreeViewOpen()).toBeFalsy()
await searchPanelModeSelector.toggleTreeView()
expect(await searchPanelModeSelector.isTreeViewOpen()).toBeTruthy()
```

If you return `!wasOpen` it could be:

```js
const isOpen = await searchPanelModeSelector.toggleTreeView()
expect(isOpen).toBeTruthy()
```
Thus, you safe two async calls and it's less code. The same applies to the other toggle functions as well.

Reply:
If we return `!wasOpen` we don't actually know if the tree-view is actually open. We just assume that the button toggled it correctly.

```
public async toggleTreeView() {
	const wasOpen = await this.isTreeViewOpen()

	await expect(page).toClick("#tree-view", { timeout: 3000 })

	if (wasOpen) {
		await page.waitForSelector("#search-panel-card", { visible: false })
	} else {
		await page.waitForSelector("#search-panel-card.expanded")
	}
        return !wasOpen
}
```

```
it("should open and close the tree-view when clicking on the tree-view icon", async () => {
	let isOpen = await searchPanelModeSelector.toggleTreeView()
	expect(isOpen).toBeTruthy()

	isOpen = await searchPanelModeSelector.toggleTreeView()
	expect(isOpen).toBeFalsy()
})
```

We can't verify that the last `await searchPanelModeSelector.toggleTreeView()`-call actually closed the tree-view, because we just return a negation of the previous state. We can't be sure about the actual state.